daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/api/Metadata.scala

```diff
@@ -0,0 +1,82 @@
+package org.apache.daffodil.runtime1.api
+
+import org.apache.daffodil.lib.exceptions.SchemaFileLocation
+import org.apache.daffodil.lib.xml.NS
+import org.apache.daffodil.lib.xml.NamedQName
+import org.apache.daffodil.runtime1.dpath.PrimTypeNode
+
+/*
+ * This is the supportable API for access to the RuntimeData structures
+ */
+trait Metadata {
+  def schemaFileLocation: SchemaFileLocation
+  def path: String
+  def diagnosticDebugName: String
+  override def toString = diagnosticDebugName
+}
+
+trait TermMetadata extends Metadata {
+  def isArray: Boolean
+}
+
+trait ElementMetadata extends TermMetadata {
+  def isSimpleType: Boolean
+  def isComplexType: Boolean
+
+  def name: String
+  def namespace: NS
+  def optNamespacePrefix: Option[String]
+  def isArray: Boolean
+  def isOptional: Boolean
+  def namedQName: NamedQName
+
+  def sscd: String
+  def isNillable: Boolean
+
+  def runtimeProperties: java.util.Map[String, String]
+
+}
+
+trait ComplexElementMetadata extends ElementMetadata {
+
+  def childMetadata: Seq[ElementMetadata]
+}
+
+trait SimpleElementMetadata extends ElementMetadata {
+  def primType: PrimTypeNode
+}
+
+trait ModelGroupMetadata extends TermMetadata {}
```

Comment:
Maybe omit these, as I have no use for them quite yet. 

Of concern in walking our metadata is that other systems may not allow metadata sharing. If not then Daffodil metadata, which shares lots of sub-structure, will blow up in size in a metadata system with unique parent pointers.